src/Google.Logging.Log4Net/LogUploader.cs

```diff
@@ -0,0 +1,154 @@
+
+        private readonly Task _uploaderTask;
+
+        private readonly AsyncAutoResetEvent _uploadReadyEvent = new AsyncAutoResetEvent();
+        private readonly AsyncAutoResetEvent _emptyEvent = new AsyncAutoResetEvent();
+
+        private DateTimeRange _logEntriesLost = null;
+
+        public void TriggerUpload(DateTimeRange logEntriesLost)
+        {
+            lock (_lockObj)
+            {
+                _logEntriesLost = _logEntriesLost != null ? _logEntriesLost.Merge(logEntriesLost) : logEntriesLost;
+                _uploadReadyEvent.Set();
+            }
+        }
+
+        private LogEntry MakeLostEntry(DateTimeRange lostRange)
+        {
+            var lostEntry = _logsLostWarningTemplate.Clone();
+            lostEntry.Timestamp = _clock.GetCurrentDateTimeUtc().ToTimestamp();
+            lostEntry.TextPayload = string.Format(lostEntry.TextPayload,
+                lostRange.From.ToString(GoogleStackdriverAppender.s_lostDateTimeFmt),
+                lostRange.To.ToString(GoogleStackdriverAppender.s_lostDateTimeFmt));
+            return lostEntry;
+        }
+
+        private async Task RunUploader()
+        {
+            TimeSpan errorDelay = _serverErrorBackoffSettings.Delay;
+            while (true)
+            {
+                IEnumerable<LogEntryExtra> entries;
+                // Wait/loop until there are some log entries that need uploading
+                while (true)
+                {
+                    DateTimeRange logEntriesLost;
+                    lock (_lockObj)
+                    {
+                        logEntriesLost = _logEntriesLost;
+                        _logEntriesLost = null;
+                    }
+                    entries = await _logQ.PeekAsync(_maxUploadBatchSize - (logEntriesLost != null ? 1 : 0));
+                    if (logEntriesLost != null)
+                    {
+                        var lostEntryExtra = new LogEntryExtra(-1, MakeLostEntry(logEntriesLost));
+                        entries = (new[] { lostEntryExtra }).Concat(entries ?? Enumerable.Empty<LogEntryExtra>());
+                    }
+                    if (entries != null && entries.Count() > 0)
+                    {
+                        // There are log entries that need uploading, so do that now.
+                        break;
+                    }
+                    await _uploadReadyEvent.WaitAsync();
+                }
+                // Upload entries to the Cloud Logging server
+                try
+                {
+                    await _client.WriteLogEntriesAsync("", null, s_emptyLabels, entries.Select(x => x.Entry));
+                    await _logQ.RemoveUntilAsync(entries.Last().Id);
+                    _emptyEvent.Set();
+                    errorDelay = _serverErrorBackoffSettings.Delay;
+                }
+                catch (Exception)
+                {
+                    // Always retry, regardless of error, as there's nothing much else to be done.
```

Comment:
If my logic is correct, this will retry for as long as there is an error on upload.
If there is an error then it waits and increases errorDelay, then goes round again to retrieve non-uploaded log entries.
Once there isn't an error on upload, it resets errorDelay.